yjit_backend.c

```diff
@@ -213,109 +213,203 @@ void ir_print_insns(jitstate_t *jit)
     }
 }
 
+void ir_print_to_dot(jitstate_t *jit)
+{
+    printf("digraph {\n");
+
+    rb_darray_for(jit->insns, insn_idx)
+    {
+        ir_insn_t insn = rb_darray_get(jit->insns, insn_idx);
+        printf("  %d [label=\"[%d] %s\"];\n", insn_idx, insn_idx, ir_op_name(insn.op));
+
+        rb_darray_for(insn.opnds, opnd_idx)
+        {
+            ir_opnd_t opnd = rb_darray_get(insn.opnds, opnd_idx);
+            if (opnd.kind == EIR_INSN_OUT) {
+                printf("  %d -> %d\n", opnd.as.idx, insn_idx);
+            }
+        }
+    }
+
+    printf("}\n");
+}
+
 /*************************************************/
 /* Convenience methods for pushing instructions. */
 /*************************************************/
 
-ir_opnd_t ir_add(jitstate_t *jit, ir_opnd_t dest, ir_opnd_t src)
+ir_opnd_t ir_add(jitstate_t *jit, ir_opnd_t opnd1, ir_opnd_t opnd2)
 {
-    return push_insn(jit, OP_ADD, dest, src);
+    return push_insn(jit, OP_ADD, opnd1, opnd2);
 }
 
-ir_opnd_t ir_mov(jitstate_t *jit, ir_opnd_t dest, ir_opnd_t src)
+ir_opnd_t ir_mov(jitstate_t *jit, ir_opnd_t opnd1, ir_opnd_t opnd2)
 {
-    return push_insn(jit, OP_MOV, dest, src);
+    RUBY_ASSERT((opnd1.kind == EIR_INSN_OUT || opnd1.kind == EIR_REG) && "can only mov into a EIR_INSN_OUT or EIR_REG");
+    return push_insn(jit, OP_MOV, opnd1, opnd2);
 }
 
 void ir_ret(jitstate_t *jit)
 {
     push_insn(jit, OP_RET);
 }
 
+ir_opnd_t ir_sub(jitstate_t *jit, ir_opnd_t opnd1, ir_opnd_t opnd2)
+{
+    return push_insn(jit, OP_SUB, opnd1, opnd2);
+}
+
 /*************************************************/
 /* Generate x86 code from the IR.                */
 /*************************************************/
 
-x86opnd_t ir_x86_opnd(insn_array_t insns, ir_opnd_t opnd)
+x86opnd_t ir_x86opnd(x86opnd_t *allocations, ir_opnd_t opnd)
 {
     switch (opnd.kind)
     {
         case EIR_REG:
             return (x86opnd_t){ OPND_REG, 64, .as.reg = { REG_GP, opnd.as.reg.idx } };
         case EIR_IMM:
             return (x86opnd_t){ OPND_IMM, opnd.num_bits, .as.imm = opnd.as.imm };
-        case EIR_INSN_OUT: {
-            // Temporary cheating way of handling register allocation that will
-            // only work for certain instructions.
-            ir_insn_t insn = rb_darray_get(insns, opnd.as.idx);
-            return ir_x86_opnd(insns, rb_darray_get(insn.opnds, 0));
-        }
+        case EIR_INSN_OUT:
+            return allocations[opnd.as.idx];
         default:
             RUBY_ASSERT(false && "unknown opnd kind");
     }
 }
 
+#define NOT_LIVE_REG -1
+
 // Fetch a free scratch register and mark it as active.
-int32_t ir_next_scratch(bool active[NUM_SCR_REGS])
+int32_t ir_next_scratch(int32_t active[NUM_SCR_REGS])
 {
     for (int32_t index = 0; index < NUM_SCR_REGS; index++) {
-        if (!active[index]) {
+        if (active[index] == NOT_LIVE_REG) {
             return index;
         }
     }
     RUBY_ASSERT(false && "out of free registers");
 }
 
```

Comment:
Where does the register marked as active in this function?

Reply:
Just asking because the comment says "mark it as active". It should probably say "find the next free scratch register" instead.